RFC: refactoring class system

dpsmith opened the pull request:
Upstream OE has several good patterns for image files some of which are abstracted into class files. Due to some of the way OpenXT is built it is not possible to use these classes directly, as a result the patterns were replicated to enable the OpenXT image recipes to look closer to their upstream siblings. 

Focus for this refactoring was:
 - eliminate duplication across recipes
 - mimic the ability to make initrafms images a dependecy for an image
 - maintain the recent move to chainable CONVERSION process
 - create a image hierarchy of  a core image and a child selinux image

Depends: PR #1221

OXT-1656

Signed-off-by: Daniel P. Smith <dpsmith@apertussolutions.com>

dpsmith commented on classes/openxt-efi.bbclass:
```diff
@@ -0,0 +1,52 @@
+		bbfatal "Unable to read GRUB_CFG"
+	fi
+
+	cat /dev/null > ${GRUB_CFG}
```
From upstream and no preference which way. The only difference is the ```cat``` will leave an empty file while ```rm``` will leave no file. I don't think that will matter but it is a subtlety that could bite if there is an expectation that the file existed later.

dpsmith commented on classes/openxt-efi.bbclass:
```diff
@@ -0,0 +1,52 @@
+	if [ -n "${APPEND}" ]; then
+		echo -e " ${APPEND}" >> ${GRUB_CFG}
+	else
+		echo "" >> ${GRUB_CFG}
```
No clue, upstream did it this way so I maintained in case of something silly grub might be doing.

dpsmith commented on classes/openxt-image-disk.bbclass:
```diff
@@ -24,87 +42,91 @@ CONVERSION_DEPENDS_disk = "syslinux \
-		done
-		chmod 0644 $BOOTDIR/initrd
-	fi
+DISK_SIGNATURE ?= "44485458"
```
Again, upstream approach. As for the significance if the string, the script just writes the bits into the field but based on endianness and field ordering in spec, it is supposed to report XTHD.

eric-ch replied:
Oh right, I did not translate the `\4\3\2\1` from the sed line.

dpsmith commented on classes/openxt-image-disk.bbclass:
```diff
@@ -24,87 +42,91 @@ CONVERSION_DEPENDS_disk = "syslinux \
 
-	# write a disk signature
-	echo -n "disk" | dd of=$IMAGE bs=1 seek=440 conv=notrunc
+	awk "BEGIN { printf \"$(echo ${DISK_SIGNATURE} | sed 's/\(..\)\(..\)\(..\)\(..\)/\\x\4\\x\3\\x\2\\x\1/')\" }" | \
```
yep, again upstream.

dpsmith commented on classes/openxt-image-disk.bbclass:
```diff
@@ -24,87 +42,91 @@ CONVERSION_DEPENDS_disk = "syslinux \
-	OFFSET=$( expr $START2 / 512 )
+validate_disk_signature() {
+	if [ $(/bin/echo -n ${DISK_SIGNATURE}|wc -c) -le 8 ]; then
+		if [ $(expr "x${DISK_SIGNATURE}" : "x[A-Za-z0-9]*$") -gt 0 ]; then
```
Upstream but if we want to rewrite in what we think is a more optimized approach, I am good with that.

eric-ch replied:
Optimized or not, this test is insufficient. Inappropriate values can get passed it.
Might worth being submitted upstream as well then.

dpsmith replied:
True, because even their own docs state that it should be a hexdigit string, not an alphnum string.

[DISK_SIGNATURE](https://www.yoctoproject.org/docs/2.3/ref-manual/ref-manual.html#var-DISK_SIGNATURE)

eric-ch replied:
Wrong on my end. `expr(1)` does parse a regular-expression and not a glob pattern for the second operand of `:`. My first test must have missed the `$` (`expr "x${DISK_SIGNATURE}" : "x[A-Za-z0-9]*"`) which made me think the second argument was not a regexp... My bad.

dpsmith commented on classes/openxt-image.bbclass:
```diff
@@ -0,0 +1,72 @@
+
+    if [ -f "${IMAGE_ROOTFS}${sysconfdir}/init.d/${script}" ]; then
+        rm -f "${IMAGE_ROOTFS}${sysconfdir}/init.d/${script}"
+        update-rc.d -r ${IMAGE_ROOTFS} ${script} remove
```
I need to double check, but I am pretty sure I copied this as is from one of the several places used. I am good with just removing the runlevel like and leaving the actual init file.

dpsmith commented on classes/openxt-image.bbclass:
```diff
@@ -0,0 +1,72 @@
+# Forcably remove packages disregarding if it creates a broken dependency
+force_package_removal() {
+    if [ -n "${PACKAGE_REMOVE}" ]; then
+        opkg -f ${IPKGCONF_TARGET} -o ${IMAGE_ROOTFS} ${OPKG_ARGS} -force-depends remove ${PACKAGE_REMOVE};
```
HA! That means it is incorrect in all the image recipes that have this. ( ^_^)

Yah, I was lazy. I just copied all the duplicate functions up into the class without applying any best practices. This would be a good time to do that, so as I will try to catch as many as possible and would welcome being doubled check on that in the next pass.

eric-ch replied:
Indeed ... :).

I have the following ones in `master`:
```
./recipes-core/images/xenclient-sysroot-image.bb:63:    opkg -f ${IPKGCONF_TARGET} -o ${IMAGE_ROOTFS} ${OPKG_ARGS} -force-depends remove ${PACKAGE_REMOVE};
./recipes-core/images/xenclient-ndvm-image.bb:92:    opkg -f ${IPKGCONF_TARGET} -o ${IMAGE_ROOTFS} ${OPKG_ARGS} -force-depends remove ${PACKAGE_REMOVE};
./recipes-core/images/xenclient-stubdomain-initramfs-image.bb:38:    opkg -f ${IPKGCONF_TARGET} -o ${IMAGE_ROOTFS} ${OPKG_ARGS} -force-depends remove ${PACKAGE_REMOVE};
```

dpsmith commented on classes/openxt-syslinux.bbclass:
```diff
@@ -0,0 +1,81 @@
+# ${SYSLINUX_MULTIBOOT_CMDLINE} - a "\n" separated list of command lines in same order as
+#	${SYSLINUX_MULTIBOOT} with empty lines for the sparse case
+#
+
```
Forgot to set expand-tab. ( ._.)

dpsmith commented on classes/openxt-vm-common.bbclass:
```diff
@@ -0,0 +1,66 @@
+	install -d $dest
+
+	# Install bzImage, initrd, and rootfs.img in DEST for all loaders to use.
+	bbnote "Trying to install ${DEPLOY_DIR_IMAGE}/${KERNEL_IMAGETYPE} as $dest/vmlinuz"
```
yep, will fix.

dpsmith commented on classes/openxt-vm-common.bbclass:
```diff
@@ -0,0 +1,66 @@
+	# Install bzImage, initrd, and rootfs.img in DEST for all loaders to use.
+	bbnote "Trying to install ${DEPLOY_DIR_IMAGE}/${KERNEL_IMAGETYPE} as $dest/vmlinuz"
+	if [ -e ${DEPLOY_DIR_IMAGE}/${KERNEL_IMAGETYPE} ]; then
+		install -m 0644 ${DEPLOY_DIR_IMAGE}/${KERNEL_IMAGETYPE} $dest/${VM_DEFAULT_KERNEL}
```
Good catch, no it will not. Will fix.

dpsmith commented on classes/openxt-vm-common.bbclass:
```diff
@@ -0,0 +1,66 @@
+		rm -f $dest/initrd
+		for fs in ${INITRD}
+		do
+			if [ -s "$fs" ]; then
```
Hmm, if we want it noted then I would prefer to do it before the for loop and dump the ```${INITRD}``` variable.

dpsmith commented on classes/openxt-efi.bbclass:
```diff
@@ -0,0 +1,52 @@
+}
+
+replace_rootfs_uuid() {
+	CFG=${1}
```
I would prefer to do a 1-for-1 in this commit. Then if everyone would like to see this reduced to the call site, handle that in a separate commit (in this PR or another).

dpsmith commented on recipes-core/images/xenclient-syncvm-image.bb:
```diff
@@ -35,24 +42,14 @@ IMAGE_INSTALL = "\
     ${@bb.utils.contains('DISTRO_FEATURES', 'blktap2', 'xen-blktap', 'blktap3', d)} \
 "
 
-require xenclient-image-common.inc
```
Yes I think it can be dropped.

dpsmith commented on classes/openxt-image.bbclass:
```diff
@@ -0,0 +1,72 @@
+}
+ROOTFS_POSTPROCESS_COMMAND += '${@bb.utils.contains("IMAGE_FEATURES", "debug-tweaks", "start_tty_on_hvc0; ", "",d)}'
+
+# Forcably remove packages disregarding if it creates a broken dependency
```
ack

dpsmith commented on recipes-core/images/xenclient-stubdomain-initramfs-image.bb:
```diff
@@ -32,11 +32,9 @@ PACKAGE_REMOVE = " \
 
 post_rootfs_shell_commands() {
-    opkg -f ${IPKGCONF_TARGET} -o ${IMAGE_ROOTFS} ${OPKG_ARGS} -force-depends remove ${PACKAGE_REMOVE};
-
```
If I understand what you are asking, this has been a ~~hack~~workaround to go behind OE's back to strip packages that are not really required/needed but were installed due to a package dependency. As a result the manifest created by OE will include the package that was stripped out and thus not align with what is in fact installed in the image.

dpsmith commented:
Applied all accepted reviewer comments.

eric-ch commented:
Build [$6679](http://openxt-builder.ainfosec.com:8010/builders/openxt64/builds/6679)
Includes:
- https://github.com/OpenXT/xenclient-oe/pull/1222
- https://github.com/OpenXT/xenclient-oe/pull/1221
- https://github.com/OpenXT/installer/pull/114

dozylynx commented:
Comparing the contents of the IMAGE_INSTALL and PACKAGE_INSTALL variables before and after applying this PR in addition to #1221 which it depends on:

dom0's IMAGE_INSTALL no longer contains 'opkg'.
dom0's PACKAGE_INSTALL has also changed: it previously contained 'opkg' twice, but now has it only once.

ndvm's IMAGE_INSTALL no longer contains 'busybox'.
ndvm's PACKAGE_INSTALL no longer contains 'busybox'.

uivm's IMAGE_INSTALL no longer contains 'opkg'.
uivm's PACKAGE_INSTALL has also changed: it previously contained 'opkg' twice, but now has it only once.

IMAGE_INSTALL and PACKAGE_INSTALL have not changed for: the installer, the stubdomain initramfs and the syncvm.

Questions raised by the above data:
* Does the NDVM filesystem still contain busybox? It seems likely that it could still be pulled in as a dependency. It would be good to confirm this.
* Are any other packages installed in the NDVM as alternative providers for eg. a shell, instead of busybox, since it is no longer an explicit package to install?
* Is opkg still installed in dom0 and the UIVM? (It should be, given its presence in PACKAGE_INSTALL.)

dpsmith commented:
> Comparing the contents of the IMAGE_INSTALL and PACKAGE_INSTALL variables before and after applying this PR in addition to #1221 which it depends on:
> 
> dom0's IMAGE_INSTALL no longer contains 'opkg'.
> dom0's PACKAGE_INSTALL has also changed: it previously contained 'opkg' twice, but now has it only once.

This double inclusion is because `${ROOTFS_PKGMANAGE}` is explicitly include into `IMAGE_INSTALL` and then included again because `package-management` is in features. We should not be explicitly including `${ROOTFS_PKGMANAGE}` in the `IMAGE_INSTALL`.

> ndvm's IMAGE_INSTALL no longer contains 'busybox'.
> ndvm's PACKAGE_INSTALL no longer contains 'busybox'.

This should not have happened. I will have to track that down.

> uivm's IMAGE_INSTALL no longer contains 'opkg'.
> uivm's PACKAGE_INSTALL has also changed: it previously contained 'opkg' twice, but now has it only once.

Same reason as why it was occuring for Dom0. 

> IMAGE_INSTALL and PACKAGE_INSTALL have not changed for: the installer, the stubdomain initramfs and the syncvm.
> 
> Questions raised by the above data:
> 
> * Does the NDVM filesystem still contain busybox? It seems likely that it could still be pulled in as a dependency. It would be good to confirm this.

It is present in my test image.

> * Are any other packages installed in the NDVM as alternative providers for eg. a shell, instead of busybox, since it is no longer an explicit package to install?

The status quo should be maintained under this PR. If there is a desire to replace/remove busybox, then it should be done under a separate PR.

> * Is opkg still installed in dom0 and the UIVM? (It should be, given its presence in PACKAGE_INSTALL.)

It is still there in Dom0 and the UIVM.

dozylynx commented:
Thanks for the answers -- this PR looks good to me.

eric-ch commented:
Thanks!
Will merge soon!

jandryuk commented on classes/openxt-image.bbclass:
```diff
@@ -0,0 +1,71 @@
+# Remove initscripts pulled-in by dependencies or not required for operation.
+remove_nonessential_initscripts() {
+    if [ -n "${INITSCRIPT_REMOVE}" ]; then
+        for i in "${INITSCRIPT_REMOVE}"; do
```
We got too aggressive adding quotes...  We need to drop them `for i in "${INITSCRIPT_REMOVE}"; do` so the for loop iterates on the space separated values.  Found this when re-basing my syncvm initscript PR.

stacktrust commented on classes/openxt-syslinux.bbclass:
```diff
@@ -0,0 +1,81 @@
+                echo -n " ---" >> ${SYSLINUX_CFG}
+            fi
+            cmdline=$(echo "$x" | sed -n ${count}p)
+            if [ -n "${cmdline}" ]; then
```
Should $x be ${SYSLINUX_MULTIBOOT_CMDLINE}?

stacktrust commented on classes/openxt-syslinux.bbclass:
```diff
@@ -0,0 +1,81 @@
+        echo -n "  APPEND" >> ${SYSLINUX_CFG}
+        count=1
+        for m in ${SYSLINUX_MULTIBOOT}; do
+            if [ ${count} -ne 1 ]; then
```
Where is count incremented?

stacktrust commented:
For ongoing regression testing of disk image classes, we may want to implement Qemu-based tests, along the lines of OE-core's meta-selftest.

https://git.openembedded.org/openembedded-core/tree/meta-selftest
